### PR TITLE
Add support for deleting (custom) metatags.

### PR DESCRIPTION
**Pull request by Marr11317**
image

### CODE REVIEW

File: mscore/metaedit.h [hunk 1]
```diff
@@ -33,21 +34,26 @@ class Score;
...
-      void setDirty() { dirty = true; }
+      void setDirty(bool dirty = true);
+      void saveClicked() { save(); } // slot for save function(which has a return value).
 
    public slots:
       virtual void accept();
 
    public:
-      MetaEditDialog(Score*, QWidget* parent = 0);
+      MetaEditDialog(Score*, QWidget *parent = nullptr);
```
Comment (Jojo-Schmitz) [inline mscore/metaedit.h]: It was just fine before, esp the position of the *

File: mscore/metaedit.h [hunk 2]
```diff
@@ -1,6 +1,7 @@
 //=============================================================================
 //  MusE Score
 //  Linux Music Score Editor
+//  $Id: metaedit.h 5290 2012-02-07 16:27:27Z wschweer $
```
Comment (Jojo-Schmitz) [inline mscore/metaedit.h]: RCS tags are not needed, please delete

File: mscore/metaedit.cpp [hunk 3]
```diff
@@ -29,95 +30,263 @@ namespace Ms {
...
-            ++idx;
+
+            QString s = i.key();
+            if (isSystemTag(s)) {
+                  QLineEdit *tag = new QLineEdit;
+                  tag->setStyleSheet("QLineEdit { background: transparent; }"); // Couldn't find any other way to remove the background.
+                  tag->setFrame(false);
+                  tag->setFocusPolicy(Qt::NoFocus);
+                  tag->setReadOnly(true);
+                  tag->setToolTip(tr("This is a system tag. It's name can't be modified."));
```
Comment (riaanvn) [inline mscore/metaedit.cpp]: This is a system tag. **Its** name can't be modified
(its = possessive; it's = "it is" contracted.)

Comment (Marr11317) [reply mscore/metaedit.cpp]: oops, sorry for that one.

File: mscore/metaedit.ui [hunk 4]
```diff
@@ -19,72 +22,178 @@
...
+     <item row="1" column="1">
+      <widget class="QLabel" name="version">
+       <property name="sizePolicy">
+        <sizepolicy hsizetype="Expanding" vsizetype="Preferred">
+         <horstretch>0</horstretch>
+         <verstretch>0</verstretch>
+        </sizepolicy>
+       </property>
+       <property name="toolTip">
+        <string>The MuseScore version which last saved the score.</string>
```
Comment (riaanvn) [inline mscore/metaedit.ui]: I was unsure about which vs that (reading some online resources confused me move. I would suggest rephrasing these in the passive voice, e.g. 
The MuseScore version with which the score was last saved.

File: mscore/metaedit.ui [hunk 5]
```diff
@@ -93,77 +202,134 @@
...
    <item>
-    <layout class="QHBoxLayout" name="horizontalLayout_2">
+    <layout class="QHBoxLayout" name="buttonLayout">
      <item>
       <widget class="QPushButton" name="newButton">
+       <property name="cursor">
+        <cursorShape>PointingHandCursor</cursorShape>
+       </property>
+       <property name="toolTip">
+        <string>Add a tag to this score.</string>
```
Comment (riaanvn) [inline mscore/metaedit.ui]: Elsewhere you refer to "the" score. I prefer "this score" to "the score" consistently, but that is a subjective preference.

File: mscore/metaedit.cpp [hunk 6]
```diff
@@ -29,95 +30,263 @@ namespace Ms {
...
-            grid->addWidget(l, idx, 0);
-            grid->addWidget(text, idx, 1);
-            ++idx;
+
+            QString s = i.key();
+            if (isSystemTag(s)) {
+                  QLineEdit *tag = new QLineEdit;
+                  tag->setStyleSheet("QLineEdit { background: transparent; }"); // Couldn't find any other way to remove the background.
+                  tag->setFrame(false);
+                  tag->setFocusPolicy(Qt::NoFocus);
```
Comment (Marr11317) [inline mscore/metaedit.cpp]: Qt::NoFocus is probably not a good idea here...

File: mscore/metaedit.cpp [hunk 7]
```diff
@@ -29,95 +30,263 @@ namespace Ms {
...
       if (rev > 99999)  // MuseScore 1.3 is decimal 5702, 2.0 and later uses a 7-digit hex SHA
             revision->setText(QString::number(rev, 16));
       else
             revision->setText(QString::number(rev, 10));
-      filePath->setText(score->importedFilePath());
-      filePath->setTextInteractionFlags(Qt::TextSelectableByMouse);
 
-      int idx = 0;
+      QString sfilePath(m_score->importedFilePath());
+      if (!sfilePath.isEmpty())
```
Comment (Marr11317) [inline mscore/metaedit.cpp]: Should I remove those unnecessary ifs? (line 45, 49, 59, one will be added on 53.)

Comment (Marr11317) [reply mscore/metaedit.cpp]: I am not sure what is the rule for that. Should we always check just in case?

File: mscore/metaedit.cpp [hunk 8]
```diff
@@ -29,96 +31,256 @@ namespace Ms {
...
+                  tag->setReadOnly(true);
+                  tag->setToolTip(tr("This is a system tag. Its name can not be modified."));
+                  tag->setText(s);
+
+                  QLineEdit *value = new QLineEdit(i.value());
+                  connect(value, SIGNAL(textChanged(const QString&)), SLOT(setDirty()));
+
+                  // If in the future the system tags can be removed,
+                  // simply uncomment the following lines. It should do the job.
+//                  QToolButton *deleteButton = new QToolButton;
```
Comment (Jojo-Schmitz) [inline mscore/metaedit.cpp]: Rather than commenting out every line, an `#if 0...#endif` wouzld IMHO be better

File: mscore/metaedit.ui (hunk 5 above)
Comment (Marr11317) [reply mscore/metaedit.ui]: Changed everything to "this score"

File: mscore/metaedit.cpp (hunk 8 above)
Comment (Marr11317) [reply mscore/metaedit.cpp]: done

**Comment by Marr11317** [comment]
Do you think existing metatags should belong to a different page? 
We could have one page for the built-in ones and one for custom ones(similar to advanced-preferences-tab).

**Comment by Jojo-Schmitz** [comment]
One screen and alphabethical order makes sense to me

**Comment by Marr11317** [comment]
But then you cant translate them...

**Comment by Jojo-Schmitz** [comment]
Why not? And why should they?

**Comment by Marr11317** [comment]
Because of user friendliness

**Comment by Marr11317** [comment]
When you edir the page header and footer youneed to understand the tags.

**Comment by Marr11317** [comment]
Ideally you should be able to change the title, lyricists, etc (I mean what is displayed on the front page) via the metaedit dialog.

**Comment by Jojo-Schmitz** [comment]
I don't see why having default and custom tags on one page prevents from making the default ones translatable

**Comment by Marr11317** [comment]
Because of the way they're made, it's tricky to translate them, then let users add their own custom metatags, then check if users are using a built-in metatag.

**Comment by Jojo-Schmitz** [comment]
The check for builtin is needed in any case, as only those can get removed.

**Comment by Marr11317** [comment]
There is no way to *untranslate* text. Which means that you want to check if the translated text, once untranslated, is equal to a built-in tag. Its actually trickier than that because you can't have twice the same tag. Anyway, you know what I mean.

**Comment by Marr11317** [comment]
I ll try a design once this one gets merged.
Do you think it's ready?

**Comment by Jojo-Schmitz** [comment]
Yes, I think it is good as it is

**Comment by Marr11317** [comment]
rebased.

**Comment by Marr11317** [comment]
Travis is not reporting that things were successful..

**Comment by Jojo-Schmitz** [comment]
Yes, the Travis integration with GitHub is currently seriously ill.

**Comment by Marr11317** [comment]
@anatoly-os A bin icon is a good idea, but I couldn't find it in Musescore.qrc 
Should I add one?

**Comment by Marr11317** [comment]
So the palettes have the bin icon (under `mscore/qml/palettes/icons`.
Should I move make a copy of it under the *normal* icon location, namely `mscore/data/icons`

**Comment by Marr11317** [comment]
Ok, done:
image

**Pull request merged**